pubsublite/internal/wire/service.go

```diff
@@ -0,0 +1,344 @@
+// error may be nil if the user called Stop().
+type serviceStatusChangeFunc func(serviceHandle, serviceStatus, error)
+
+// service is the interface that must be implemented by services (essentially
+// gRPC client stream wrappers, e.g. subscriber, publisher) that can be
+// dependencies of a compositeService.
+type service interface {
+	Start()
+	Stop()
+
+	// Methods below are implemented by abstractService.
+	AddStatusChangeReceiver(serviceHandle, serviceStatusChangeFunc)
+	RemoveStatusChangeReceiver(serviceHandle)
+	Handle() serviceHandle
+	Error() error
+}
+
+// abstractService can be embedded into other structs to provide common
+// functionality for managing service status and status change receivers.
+type abstractService struct {
+	mu sync.Mutex
+
+	statusChangeReceivers []*statusChangeReceiver
+	status                serviceStatus
+	// The error that cause the service to terminate.
+	err error
+}
+
+type statusChangeReceiver struct {
+	handle         serviceHandle // For removing the receiver.
+	onStatusChange serviceStatusChangeFunc
+}
+
+func (as *abstractService) AddStatusChangeReceiver(handle serviceHandle, onStatusChange serviceStatusChangeFunc) {
+	as.mu.Lock()
+	defer as.mu.Unlock()
+	as.statusChangeReceivers = append(
+		as.statusChangeReceivers,
+		&statusChangeReceiver{handle, onStatusChange})
+}
+
+func (as *abstractService) RemoveStatusChangeReceiver(handle serviceHandle) {
+	as.mu.Lock()
+	defer as.mu.Unlock()
+
+	for i := len(as.statusChangeReceivers) - 1; i >= 0; i-- {
+		r := as.statusChangeReceivers[i]
+		if r.handle == handle {
+			// Swap with last element, erase last element and truncate the slice.
+			lastIdx := len(as.statusChangeReceivers) - 1
+			if i != lastIdx {
+				as.statusChangeReceivers[i] = as.statusChangeReceivers[lastIdx]
+			}
+			as.statusChangeReceivers[lastIdx] = nil
+			as.statusChangeReceivers = as.statusChangeReceivers[:lastIdx]
+		}
+	}
+}
+
+// Handle identifies this service instance, even when there are multiple layers
+// of embedding.
+func (as *abstractService) Handle() serviceHandle {
+	return as
+}
+
+func (as *abstractService) Error() error {
+	as.mu.Lock()
+	defer as.mu.Unlock()
+	return as.err
+}
+
+func (as *abstractService) Status() serviceStatus {
+	as.mu.Lock()
+	defer as.mu.Unlock()
+	return as.status
+}
+
+func (as *abstractService) unsafeCheckServiceStatus() error {
+	switch {
+	case as.status == serviceUninitialized:
+		return ErrServiceUninitialized
+	case as.status == serviceStarting:
+		return ErrServiceStarting
+	case as.status >= serviceTerminating:
+		return ErrServiceStopped
+	default:
+		return nil
+	}
+}
+
+// unsafeUpdateStatus assumes the service is already holding a mutex when
+// called, as it often needs to be atomic with other operations.
+func (as *abstractService) unsafeUpdateStatus(targetStatus serviceStatus, err error) bool {
+	if as.status >= targetStatus {
+		// Already at the same or later stage of the service lifecycle.
+		return false
```

Comment:
Should `as.err` not be updated in this situation?

Reply:
Yes, the errors for the obsolete statuses are ignored (only `as.status == targetStatus` actually occurs). Errors are really only set for serviceTerminating or serviceTerminated and we try not to clobber the first error (below).